services/agent/src/simcore_service_agent/services/docker_utils.py

```diff
@@ -106,3 +106,27 @@ async def remove_volume(
         get_instrumentation(app).agent_metrics.remove_volumes(
             settings.AGENT_DOCKER_NODE_ID
         )
+
+
+async def get_containers_with_prefixes(docker: Docker, prefixes: set[str]) -> set[str]:
+    """Returns a set of container names matching any of the given prefixes"""
+    all_containers = await docker.containers.list(all=True)
+
+    result: set[str] = set()
+    for container in all_containers:
+        container_info = await container.show()
+        container_name = container_info.get("Name", "").lstrip("/")
+        if any(container_name.startswith(prefix) for prefix in prefixes):
+            result.add(container_name)
+
+    return result
+
+
+async def remove_container_forcefully(docker: Docker, container_id: str) -> None:
+    """Removes a container regardless of it's state"""
+    try:
+        container = await docker.containers.get(container_id)
+        await container.delete(force=True)
```

Comment:
Raise a warning as a log here, please. It should afterall, in theory, never happen :D

Reply:
No need here. It will create noise and confusion. Already thought about that and you have a clear message that points out if something did not work as expected. Those would be the `unexpected_orphans` message.

Reply:
@GitHK change request here

Reply:
@GitHK rebuttal

Reply:
Walk with me though the issue, I really do not get the point of adding a debug line here.


## In case of error

To be precise, when aiodocker raises an error.

We try to remove the container and two things cn happen:
1. Docker replies with 404 (not found), as in the container was not found. Since the container was not found it's fine not to do anything here.
2. If the status is different, the error is raised and properly reported.


If your question is: why can a 404 happen? The answer is:
- `concurrency`: a container can be listed as existing, then when you try to remove it it can have already disappeared. **NOTHING** can exclude this from never happening!


## In case there are no errors

This is the expected behaviour and the container is removed. We don't need to log anything more here.

Reply:
A lock can exclude the container being deleted while it is in the state of being deleted. The director-v2 is free to create some DB entry somewhere, or anything like a lock, that will inform `agents` or any other microservices that the service, although still listed as existing wen using the DockerAPI, is already scheduled to be deleted, at a given time. Just to give an example and to rebuttal your statement "*NOTHING* can exclude this from happening". I think a lock can prevent this from happening.

Now, back to the topic. If the agent removes an orphaned container, I want there to be a warning logline that informs us that there was an orpahend container. There should not ever be an orphaned container, do you agree? Why do you propose to have this action silent/silenced? @GitHK

Reply:
I think what you are asking was already done in my above comment, where I have already aggregated all the errors.

---

To clear up the `**NOTHING** can exclude this from happening`. This affirmation  still stands. If a service is removed, its containers take a few seconds to be removed from the services.
In the meanwhile, you can still list them. The issue is that docker will remove it and you cannot block this operation from happening. I was thinking about this situation.